### DISCUSSION TITLE
[SUGGESTION] Vertical SubClaiming

### DISCUSSION THREAD

**Discussion by ShaneBeee** [GriefPrevention/GriefPrevention#269] [Ideas]
Hello
I would like to suggest vertical subclaims. Example, being able to claim say a 10x10 area, but also with a Y value of say 10. This way you could stack subclaims, like if you were doing condos/apartments in your little town.

I personally would LOVE to have this feature on my server.
Thanks for reading!

**Comment by bigpresh** [comment]
This has been suggested in the past, and I believe it was rejected because it would be too expensive to check.

It *might* be less expensive if "stackability" only applied to subclaims, rather than top-level claims, perhaps.

I'm not sure how creating and visualising stacked claims could be done in a non-confusing manner.  Maybe a command that limits the height of a subclaim to be created, e.g. `/subclaimheight 10`, which you'd use before creating it, and it would treat the claim as being from the Y-value of the block you right click first to that many blocks above it.

I definitely see the appeal, though!

**Reply by Kycermann** [reply]
Could we not use octrees inside of claims? They have log2 lookup complexity- and the complexity for writing doesn't really matter because people hardly ever modify claims compared to how much they're read.

**Reply by joex92** [reply]
> I'm not sure how creating and visualising stacked claims could be done in a non-confusing manner. Maybe a command that limits the height of a subclaim to be created, e.g. `/subclaimheight 10`, which you'd use before creating it, and it would treat the claim as being from the Y-value of the block you right click first to that many blocks above it.

I believe the best approach would be an extra command or a sub command for the subclaim command, like this:
`/3dsubdivideclaim` or `/subdivideclaim 3d`
Either way you'll have to select 2 corners like the worldedit selection tool...

**Comment by ShaneBeee** [comment]
Well... I totally thought I replied to this. Yeah I was referring STRICTLY to subclaims. As stacking top level claims would make absolutely no sense. 
I Feel like I have been in a few situations where vertical sub claiming would be handy dandy. Like I have a room in my house I want to give someone Perm to, but I dont want them to be able to access the chests on the lower floor. 
And I know a lot of my players would love this, especially since I have the GP Real Estate addon.... would be able to do a lot more.
If I knew how to write java, I would be ALL over this ;)

**Comment by LadyViviann** [comment]
This would be very handy for the crew building a subway system on our server as well where people would like to give perms to the builders, but if the perms are taken away after the build and someone gets out of their minecart, they can't get back in it again. Or they are left with access to the entire build.

**Comment by TomLewis** [comment]
If you are doing rentable condos/apartments, you can make it work with areashop to rent vertical regions using worldguard. I use this for my shops.

**Comment by Weasel-Beans** [comment]
This would definitely be helpful for what I'm doing with claims in the spawn area of my server

**Reply by iSyriux2** [reply]
Yeah same I had this floating desert base in a towny server and I had a plan for vertical apartments but then I found out that it took the entire chunk, so if I had a public facility under someone's private plot, they could come and grief it

**Comment by joex92** [comment]
So any follow up on this? is it going to be implemented? or at least is it planned?

**Reply by Jikoo** [reply]
This is not planned. It is planned to be supported via an addon, but it is also not a near future thing because it requires a rewrite of the datastore system. You can see the roadmap in #1162, but bear in mind that this is an open source project maintained by people who aren't paid to do so at all, so progress is slow.

**Comment by castledking** [comment]
https://github.com/castledking/GriefPrevention/releases/

right click to corners of a cube >2 blocks in height while in /subdivideclaims mode. trust somebody inside that subdivision. enjoy.

**Reply by zefir-git** [reply]
have you considered opening a PR?

**Reply by castledking** [reply]
it probably wont get merged into the main plugin, the current dev has stated that adding 3D subdivisions is not in his scope of making an anti-grief plugin. It's pretty stable though, with some production servers using it bug-free for the past month, I'll try a pull request sometime later.

**Reply by RoboMWM** [reply]
yea, not in scope, but very much welcome as an addon (if it's possible? If not, then would like to make the API as such for at least v20 to allow for this!)

I want addons to have full control on how subclaims are managed, as there are a lot of different use cases and features requested for how subclaims behave.

**Reply by Jikoo** [reply]
I think for it to work as an addon it would have to replace at least the player interact listener and possibly the datastore. Possible, yes. Possibly painful though. Probably less painful than maintaining a whole fork of GP though.
